code/openvpn_v2.6.13_develop/doc/openvpn.8
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "OPENVPN" 8 "" "" "System Manager's Manual"
.SH NAME
openvpn \- Secure IP tunnel daemon
.SH SYNOPSIS
.nf
\fBopenvpn\fP [ options ... ]
\fBopenvpn\fP  \fB\-\-help\fP
.fi
.sp
.SH INTRODUCTION
.sp
OpenVPN is an open source VPN daemon by James Yonan. Because OpenVPN
tries to be a universal VPN tool offering a great deal of flexibility,
there are a lot of options on this manual page. If you\(aqre new to
OpenVPN, you might want to skip ahead to the examples section where you
will see how to construct simple VPNs on the command line without even
needing a configuration file.
.sp
Also note that there\(aqs more documentation and examples on the OpenVPN
web site: \fI\%https://openvpn.net/\fP
.sp
And if you would like to see a shorter version of this manual, see the
openvpn usage message which can be obtained by running \fBopenvpn\fP
without any parameters.
.SH DESCRIPTION
.sp
OpenVPN is a robust and highly flexible VPN daemon. OpenVPN supports
SSL/TLS security, ethernet bridging, TCP or UDP tunnel transport through
proxies or NAT, support for dynamic IP addresses and DHCP, scalability
to hundreds or thousands of users, and portability to most major OS
platforms.
.sp
OpenVPN is tightly bound to the OpenSSL library, and derives much of its
crypto capabilities from it.
.sp
OpenVPN supports conventional encryption using a pre\-shared secret key
\fB(Static Key mode)\fP or public key security \fB(SSL/TLS mode)\fP using
client & server certificates. OpenVPN also supports non\-encrypted
TCP/UDP tunnels.
.sp
OpenVPN is designed to work with the \fBTUN/TAP\fP virtual networking
interface that exists on most platforms.
.sp
Overall, OpenVPN aims to offer many of the key features of IPSec but
with a relatively lightweight footprint.
.SH OPTIONS
.sp
OpenVPN allows any option to be placed either on the command line or in
a configuration file. Though all command line options are preceded by a
double\-leading\-dash ("\-\-"), this prefix can be removed when an option is
placed in a configuration file.
.SS Generic Options
.sp
This section covers generic options which are accessible regardless of
which mode OpenVPN is configured as.
.INDENT 0.0
.TP
.B  \-\-help
Show options.
.TP
.B  \-\-auth\-nocache
Don\(aqt cache \fB\-\-askpass\fP or \fB\-\-auth\-user\-pass\fP username/passwords in
virtual memory.
.sp
If specified, this directive will cause OpenVPN to immediately forget
username/password inputs after they are used. As a result, when OpenVPN
needs a username/password, it will prompt for input from stdin, which
may be multiple times during the duration of an OpenVPN session.
.sp
When using \fB\-\-auth\-nocache\fP in combination with a user/password file
and \fB\-\-chroot\fP or \fB\-\-daemon\fP, make sure to use an absolute path.
.TP
.BI \-\-cd \ dir
Change directory to \fBdir\fP prior to reading any files such as
configuration files, key files, scripts, etc. \fBdir\fP should be an
absolute path, with a leading "/", and without any references to the
current directory such as \fB\&.\fP or \fB\&..\fP\&.
.sp
This option is useful when you are running OpenVPN in \fB\-\-daemon\fP mode,
and you want to consolidate all of your OpenVPN control files in one
location.
.TP
.BI \-\-chroot \ dir
Chroot to \fBdir\fP after initialization. \fB\-\-chroot\fP essentially
redefines \fBdir\fP as being the top level directory tree (/). OpenVPN
will therefore be unable to access any files outside this tree. This can
be desirable from a security standpoint.
.sp
Since the chroot operation is delayed until after initialization, most
OpenVPN options that reference files will operate in a pre\-chroot
context.
.sp
In many cases, the \fBdir\fP parameter can point to an empty directory,
however complications can result when scripts or restarts are executed
after the chroot operation.
.sp
Note: The SSL library will probably need /dev/urandom to be available
inside the chroot directory \fBdir\fP\&. This is because SSL libraries
occasionally need to collect fresh randomness. Newer linux kernels and some
BSDs implement a getrandom() or getentropy() syscall that removes the
need for /dev/urandom to be available.
.TP
.BI \-\-compat\-mode \ version
This option provides a convenient way to alter the defaults of OpenVPN
to be more compatible with the version \fBversion\fP specified. All of
the changes this option applies can also be achieved using individual
configuration options.
.sp
The version specified with this option is the version of OpenVPN peer
OpenVPN should try to be compatible with. In general OpenVPN should be
compatible with the last two previous version without this option. E.g.
OpenVPN 2.6.0 should be compatible with 2.5.x and 2.4.x without this option.
However, there might be some edge cases that still require this option even
in these cases.
.sp
Note: Using this option reverts defaults to no longer recommended
values and should be avoided if possible.
.sp
The following table details what defaults are changed depending on the
version specified.
.INDENT 7.0
.IP \(bu 2
2.5.x or lower: \fB\-\-allow\-compression asym\fP is automatically added
to the configuration if no other compression options are present.
.IP \(bu 2
2.4.x or lower: The cipher in \fB\-\-cipher\fP is appended to
\fB\-\-data\-ciphers\fP\&.
.IP \(bu 2
2.3.x or lower: \fB\-\-data\-ciphers\-fallback\fP is automatically added with
the same cipher as \fB\-\-cipher\fP\&.
.IP \(bu 2
2.3.6 or lower: \fB\-\-tls\-version\-min 1.0\fP is added to the configuration
when \fB\-\-tls\-version\-min\fP is not explicitly set.
.UNINDENT
.sp
If not required, this is option should be avoided. Setting this option can
lower security or disable features like data\-channel offloading.
.TP
.BI \-\-config \ file
Load additional config options from \fBfile\fP where each line corresponds
to one command line option, but with the leading \fB\-\-\fP removed.
.sp
If \fB\-\-config file\fP is the only option to the openvpn command, the
\fB\-\-config\fP can be removed, and the command can be given as \fBopenvpn
file\fP
.sp
Note that configuration files can be nested to a reasonable depth.
.sp
Double quotation or single quotation characters ("", \(aq\(aq) can be used to
enclose single parameters containing whitespace, and "#" or ";"
characters in the first column can be used to denote comments.
.sp
Note that OpenVPN 2.0 and higher performs backslash\-based shell escaping
for characters not in single quotations, so the following mappings
should be observed:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\e\e       Maps to a single backslash character (\e).
\e"       Pass a literal doublequote character ("), don\(aqt
         interpret it as enclosing a parameter.
\e[SPACE] Pass a literal space or tab character, don\(aqt
         interpret it as a parameter delimiter.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For example on Windows, use double backslashes to represent pathnames:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
secret "c:\e\eOpenVPN\e\esecret.key"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For examples of configuration files, see
\fI\%https://openvpn.net/community\-resources/how\-to/\fP
.sp
Here is an example configuration file:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
#
# Sample OpenVPN configuration file for
# using a pre\-shared static key.
#
# \(aq#\(aq or \(aq;\(aq may be used to delimit comments.

# Use a dynamic tun device.
dev tun

# Our remote peer
remote mypeer.mydomain

# 10.1.0.1 is our local VPN endpoint
# 10.1.0.2 is our remote VPN endpoint
ifconfig 10.1.0.1 10.1.0.2

# Our pre\-shared static key
secret static.key
.ft P
.fi
.UNINDENT
.UNINDENT
.TP
.BI \-\-daemon \ progname
Become a daemon after all initialization functions are completed.
.sp
Valid syntaxes:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
daemon
daemon progname
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This option will cause all message and error output to be sent to the syslog
file (such as \fB/var/log/messages\fP), except for the output of
scripts and ifconfig commands, which will go to \fB/dev/null\fP unless
otherwise redirected. The syslog redirection occurs immediately at the
point that \fB\-\-daemon\fP is parsed on the command line even though the
daemonization point occurs later. If one of the \fB\-\-log\fP options is
present, it will supersede syslog redirection.
.sp
The optional \fBprogname\fP parameter will cause OpenVPN to report its
program name to the system logger as \fBprogname\fP\&. This can be useful in
linking OpenVPN messages in the syslog file with specific tunnels. When
unspecified, \fBprogname\fP defaults to \fBopenvpn\fP\&.
.sp
When OpenVPN is run with the \fB\-\-daemon\fP option, it will try to delay
daemonization until the majority of initialization functions which are
capable of generating fatal errors are complete. This means that
initialization scripts can test the return status of the openvpn command
for a fairly reliable indication of whether the command has correctly
initialized and entered the packet forwarding event loop.
.sp
In OpenVPN, the vast majority of errors which occur after initialization
are non\-fatal.
.sp
Note: as soon as OpenVPN has daemonized, it can not ask for usernames,
passwords, or key pass phrases anymore. This has certain consequences,
namely that using a password\-protected private key will fail unless the
\fB\-\-askpass\fP option is used to tell OpenVPN to ask for the pass phrase
(this requirement is new in v2.3.7, and is a consequence of calling
daemon() before initializing the crypto layer).
.sp
Further, using \fB\-\-daemon\fP together with \fB\-\-auth\-user\-pass\fP (entered
on console) and \fB\-\-auth\-nocache\fP will fail as soon as key
renegotiation (and reauthentication) occurs.
.TP
.B  \-\-disable\-dco
Disable "data channel offload" (DCO).
.sp
On Linux don\(aqt use the ovpn\-dco device driver, but rather rely on the
legacy tun module.
.sp
You may want to use this option if your server needs to allow clients
older than version 2.4 to connect.
.TP
.B  \-\-disable\-occ
\fBDEPRECATED\fP Disable "options consistency check" (OCC) in configurations
that do not use TLS.
.sp
Don\(aqt output a warning message if option inconsistencies are detected
between peers. An example of an option inconsistency would be where one
peer uses \fB\-\-dev tun\fP while the other peer uses \fB\-\-dev tap\fP\&.
.sp
Use of this option is discouraged, but is provided as a temporary fix in
situations where a recent version of OpenVPN must connect to an old
version.
.TP
.BI \-\-engine \ engine\-name
Enable OpenSSL hardware\-based crypto engine functionality.
.sp
Valid syntaxes:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
engine
engine engine\-name
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If \fBengine\-name\fP is specified, use a specific crypto engine. Use the
\fB\-\-show\-engines\fP standalone option to list the crypto engines which
are supported by OpenSSL.
.TP
.B  \-\-fast\-io
Optimize TUN/TAP/UDP I/O writes by avoiding a call to
poll/epoll/select prior to the write operation. The purpose of such a
call would normally be to block until the device or socket is ready to
accept the write. Such blocking is unnecessary on some platforms which
don\(aqt support write blocking on UDP sockets or TUN/TAP devices. In such
cases, one can optimize the event loop by avoiding the poll/epoll/select
call, improving CPU efficiency by 5% to 10%.
.sp
This option can only be used on non\-Windows systems, when \fB\-\-proto
udp\fP is specified, and when \fB\-\-shaper\fP is \fINOT\fP specified.
.TP
.BI \-\-group \ group
Similar to the \fB\-\-user\fP option, this option changes the group ID of
the OpenVPN process to \fBgroup\fP after initialization.
.TP
.BI \-\-ignore\-unknown\-option \ args
Valid syntax:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
ignore\-unknown\-options opt1 opt2 opt3 ... optN
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
When one of options \fBopt1 ... optN\fP is encountered in the configuration
file the configuration file parsing does not fail if this OpenVPN version
does not support the option. Multiple \fB\-\-ignore\-unknown\-option\fP options
can be given to support a larger number of options to ignore.
.sp
This option should be used with caution, as there are good security
reasons for having OpenVPN fail if it detects problems in a config file.
Having said that, there are valid reasons for wanting new software
features to gracefully degrade when encountered by older software
versions.
.sp
\fB\-\-ignore\-unknown\-option\fP is available since OpenVPN 2.3.3.
.TP
.BI \-\-iproute \ cmd
Set alternate command to execute instead of default \fBiproute2\fP command.
May be used in order to execute OpenVPN in unprivileged environment.
.TP
.BI \-\-keying\-material\-exporter \ args
Save Exported Keying Material [RFC5705] of \fBlen\fP bytes (must be between 16
and 4095 bytes) using \fBlabel\fP in environment
(\fBexported_keying_material\fP) for use by plugins in
\fBOPENVPN_PLUGIN_TLS_FINAL\fP callback.
.sp
Valid syntax:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
keying\-material\-exporter label len
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Note that exporter \fBlabels\fP have the potential to collide with existing
PRF labels. In order to prevent this, labels \fIMUST\fP begin with
\fBEXPORTER\fP\&.
.TP
.B  \-\-mlock
Disable paging by calling the POSIX mlockall function. Requires that
OpenVPN be initially run as root (though OpenVPN can subsequently
downgrade its UID using the \fB\-\-user\fP option).
.sp
Using this option ensures that key material and tunnel data are never
written to disk due to virtual memory paging operations which occur
under most modern operating systems. It ensures that even if an attacker
was able to crack the box running OpenVPN, he would not be able to scan
the system swap file to recover previously used ephemeral keys, which
are used for a period of time governed by the \fB\-\-reneg\fP options (see
below), then are discarded.
.sp
The downside of using \fB\-\-mlock\fP is that it will reduce the amount of
physical memory available to other applications.
.sp
The limit on how much memory can be locked and how that limit
is enforced are OS\-dependent. On Linux the default limit that an
unprivileged process may lock (RLIMIT_MEMLOCK) is low, and if
privileges are dropped later, future memory allocations will very
likely fail. The limit can be increased using ulimit or systemd
directives depending on how OpenVPN is started.
.sp
If the platform has the getrlimit(2) system call, OpenVPN will check
for the amount of mlock\-able memory before calling mlockall(2), and
tries to increase the limit to 100 MB if less than this is configured.
100 Mb is somewhat arbitrary \- it is enough for a moderately\-sized
OpenVPN deployment, but the memory usage might go beyond that if the
number of concurrent clients is high.
.TP
.BI \-\-nice \ n
Change process priority after initialization (\fBn\fP greater than 0 is
lower priority, \fBn\fP less than zero is higher priority).
.TP
.BI \-\-providers \ providers
Load the list of (OpenSSL) providers. This is mainly useful for using an
external provider for key management like tpm2\-openssl or to load the
legacy provider with
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-\-providers legacy default
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Behaviour of changing this option between \fBSIGHUP\fP might not be well behaving.
If you need to change/add/remove this option, fully restart OpenVPN.
.TP
.BI \-\-remap\-usr1 \ signal
Control whether internally or externally generated \fBSIGUSR1\fP signals
are remapped to \fBSIGHUP\fP (restart without persisting state) or
\fBSIGTERM\fP (exit).
.sp
\fBsignal\fP can be set to \fBSIGHUP\fP or \fBSIGTERM\fP\&. By default,
no remapping occurs.
.TP
.BI \-\-script\-security \ level
This directive offers policy\-level control over OpenVPN\(aqs usage of
external programs and scripts. Lower \fBlevel\fP values are more
restrictive, higher values are more permissive. Settings for \fBlevel\fP:
.INDENT 7.0
.TP
.B \fB0\fP
Strictly no calling of external programs.
.TP
.B \fB1\fP
(Default) Only call built\-in executables such as ifconfig,
ip, route, or netsh.
.TP
.B \fB2\fP
Allow calling of built\-in executables and user\-defined
scripts.
.TP
.B \fB3\fP
Allow passwords to be passed to scripts via environmental
variables (potentially unsafe).
.UNINDENT
.sp
OpenVPN releases before v2.3 also supported a \fBmethod\fP flag which
indicated how OpenVPN should call external commands and scripts. This
could be either \fBexecve\fP or \fBsystem\fP\&. As of OpenVPN 2.3, this
flag is no longer accepted. In most *nix environments the execve()
approach has been used without any issues.
.sp
Some directives such as \fB\-\-up\fP allow options to be passed to the
external script. In these cases make sure the script name does not
contain any spaces or the configuration parser will choke because it
can\(aqt determine where the script name ends and script options start.
.sp
To run scripts in Windows in earlier OpenVPN versions you needed to
either add a full path to the script interpreter which can parse the
script or use the \fBsystem\fP flag to run these scripts. As of OpenVPN
2.3 it is now a strict requirement to have full path to the script
interpreter when running non\-executables files. This is not needed for
executable files, such as .exe, .com, .bat or .cmd files. For example,
if you have a Visual Basic script, you must use this syntax now:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-\-up \(aqC:\e\eWindows\e\eSystem32\e\ewscript.exe C:\e\eProgram\e Files\e\eOpenVPN\e\econfig\e\emy\-up\-script.vbs\(aq
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Please note the single quote marks and the escaping of the backslashes
(\e) and the space character.
.sp
The reason the support for the \fBsystem\fP flag was removed is due to
the security implications with shell expansions when executing scripts
via the \fBsystem()\fP call.
.TP
.BI \-\-setcon \ context
Apply SELinux \fBcontext\fP after initialization. This essentially
provides the ability to restrict OpenVPN\(aqs rights to only network I/O
operations, thanks to SELinux. This goes further than \fB\-\-user\fP and
\fB\-\-chroot\fP in that those two, while being great security features,
unfortunately do not protect against privilege escalation by
exploitation of a vulnerable system call. You can of course combine all
three, but please note that since setcon requires access to /proc you
will have to provide it inside the chroot directory (e.g. with mount
\-\-bind).
.sp
Since the setcon operation is delayed until after initialization,
OpenVPN can be restricted to just network\-related system calls, whereas
by applying the context before startup (such as the OpenVPN one provided
in the SELinux Reference Policies) you will have to allow many things
required only during initialization.
.sp
Like with chroot, complications can result when scripts or restarts are
executed after the setcon operation, which is why you should really
consider using the \fB\-\-persist\-tun\fP option.
.TP
.BI \-\-status \ args
Write operational status to \fBfile\fP every \fBn\fP seconds. \fBn\fP defaults
to \fB60\fP if not specified.
.sp
Valid syntaxes:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
status file
status file n
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Status can also be written to the syslog by sending a \fBSIGUSR2\fP
signal.
.sp
With multi\-client capability enabled on a server, the status file
includes a list of clients and a routing table. The output format can be
controlled by the \fB\-\-status\-version\fP option in that case.
.sp
For clients or instances running in point\-to\-point mode, it will contain
the traffic statistics.
.TP
.BI \-\-status\-version \ n
Set the status file format version number to \fBn\fP\&.
.sp
This only affects the status file on servers with multi\-client
capability enabled.  Valid status version values:
.INDENT 7.0
.TP
.B \fB1\fP
Traditional format (default). The client list contains the
following fields comma\-separated: Common Name, Real Address, Bytes
Received, Bytes Sent, Connected Since.
.TP
.B \fB2\fP
A more reliable format for external processing. Compared to
version \fB1\fP, the client list contains some additional fields:
Virtual Address, Virtual IPv6 Address, Username, Client ID, Peer ID,
Data Channel Cipher. Future versions may extend the number of fields.
.TP
.B \fB3\fP
Identical to \fB2\fP, but fields are tab\-separated.
.UNINDENT
.TP
.B  \-\-test\-crypto
Do a self\-test of OpenVPN\(aqs crypto options by encrypting and decrypting
test packets using the data channel encryption options specified above.
This option does not require a peer to function, and therefore can be
specified without \fB\-\-dev\fP or \fB\-\-remote\fP\&.
.sp
The typical usage of \fB\-\-test\-crypto\fP would be something like this:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
openvpn \-\-test\-crypto \-\-secret key
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
or
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
openvpn \-\-test\-crypto \-\-secret key \-\-verb 9
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This option is very useful to test OpenVPN after it has been ported to a
new platform, or to isolate problems in the compiler, OpenSSL crypto
library, or OpenVPN\(aqs crypto code. Since it is a self\-test mode,
problems with encryption and authentication can be debugged
independently of network and tunnel issues.
.TP
.BI \-\-tmp\-dir \ dir
Specify a directory \fBdir\fP for temporary files. This directory will be
used by openvpn processes and script to communicate temporary data with
openvpn main process. Note that the directory must be writable by the
OpenVPN process after it has dropped it\(aqs root privileges.
.sp
This directory will be used by in the following cases:
.INDENT 7.0
.IP \(bu 2
\fB\-\-client\-connect\fP scripts and \fBOPENVPN_PLUGIN_CLIENT_CONNECT\fP
plug\-in hook to dynamically generate client\-specific configuration
\fBclient_connect_config_file\fP and return success/failure via
\fBclient_connect_deferred_file\fP when using deferred client connect
method
.IP \(bu 2
\fBOPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY\fP plug\-in hooks returns
success/failure via \fBauth_control_file\fP when using deferred auth
method and pending authentication via \fBpending_auth_file\fP\&.
.UNINDENT
.TP
.B  \-\-use\-prediction\-resistance
Enable prediction resistance on mbed TLS\(aqs RNG.
.sp
Enabling prediction resistance causes the RNG to reseed in each call for
random. Reseeding this often can quickly deplete the kernel entropy
pool.
.sp
If you need this option, please consider running a daemon that adds
entropy to the kernel pool.
.TP
.BI \-\-user \ user
Change the user ID of the OpenVPN process to \fBuser\fP after
initialization, dropping privileges in the process. This option is
useful to protect the system in the event that some hostile party was

code/openvpn_v2.6.13_develop/src/openvpn/openvpn.c
/*
 *  OpenVPN -- An application to securely tunnel IP networks
 *             over a single TCP/UDP port, with support for SSL/TLS-based
 *             session authentication and key exchange,
 *             packet encryption, packet authentication, and
 *             packet compression.
 *
 *  Copyright (C) 2002-2024 OpenVPN Inc <sales@openvpn.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 2
 *  as published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/*
The #ifdef ... #endif block is a preprocessor directive used in C and C++ to conditionally include or exclude code during compilation.
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "syshead.h"

#include "init.h"
#include "forward.h"
#include "multi.h"
#include "win32.h"
#include "platform.h"
#include "string.h"

#include "memdbg.h"

#define P2P_CHECK_SIG() EVENT_LOOP_CHECK_SIGNAL(c, process_signal_p2p, c);

static bool
process_signal_p2p(struct context *c)
{
    remap_signal(c);
    return process_signal(c);
}

/**************************************************************************/
/**
 * Main event loop for OpenVPN in client mode, where only one VPN tunnel
 * is active.
 * @ingroup eventloop
 *
 * @param c - The context structure of the single active VPN tunnel.
 */
static void

/*The tunnel_point_to_point function sets up and manages the VPN tunnel's state, prepares for I/O, and processes data packets.*/

/*msg(M_INFO, "%s", "==========in to init crypto layer");*/

tunnel_point_to_point(struct context *c)
{
    context_clear_2(c);

    /* set point-to-point mode */
    c->mode = CM_P2P;
    /* initialize tunnel instance, avoid SIGHUP when config is stdin since
     * re-reading the config from stdin will not work */
    bool stdin_config = c->options.config && (strcmp(c->options.config, "stdin") == 0);

    /*The function init_instance_handle_signals is responsible for setting up the tunnel instance, configuring it (including mutual authentication, key exchange, and tunnel setup), and managing signal handling both before and after the initialization of the tunnel.*/

    init_instance_handle_signals(c, c->es, stdin_config ? 0 : CC_HARD_USR1_TO_HUP);
    if (IS_SIG(c))
    {
        return;
    }

    /* main event loop */
    while (true)
    {
        perf_push(PERF_EVENT_LOOP);

        /* process timers, TLS, etc. */
        // msg(M_INFO, "========== Begining of pre_select %s", c->options.ciphername);
        // msg(M_INFO, "========== Begining of pre_select %s", c->options.ncp_ciphers);
        pre_select(c);
        // msg(M_INFO, "========== End of pre_select %s", c->options.ciphername);
        // msg(M_INFO, "========== End of pre_select %s", c->options.ncp_ciphers);
        P2P_CHECK_SIG();

        /* set up and do the I/O wait */
        io_wait(c, p2p_iow_flags(c));
        P2P_CHECK_SIG();

        /* timeout? */
        if (c->c2.event_set_status == ES_TIMEOUT)
        {
            perf_pop();
            continue;
        }

        /* process the I/O which triggered select */
        // msg(M_INFO, "========== Begining of process_io %s", c->options.ciphername);
        // msg(M_INFO, "========== Begining of process_io %s", c->options.ncp_ciphers);
        process_io(c, c->c2.link_sockets[0]);
        P2P_CHECK_SIG();

        perf_pop();
    }

    persist_client_stats(c);

    uninit_management_callback();

    /* tear down tunnel instance (unless --persist-tun) */
    close_instance(c);
}

#undef PROCESS_SIGNAL_P2P

/*init_early: Initializes key components like the network context, verbosity levels, and cryptographic providers early in the process.*/

void init_early(struct context *c)
{
    net_ctx_init(c, &c->net_ctx);

    /* init verbosity and mute levels */
    init_verb_mute(c, IVM_LEVEL_1);

    /* Initialise OpenSSL provider, this needs to be initialised this
     * early since option post-processing and also openssl info
     * printing depends on it */
    for (int j = 1; j < MAX_PARMS && c->options.providers.names[j]; j++)
    {
        c->options.providers.providers[j] =
            crypto_load_provider(c->options.providers.names[j]);
    }
}

/*uninit_early: Frees the resources initialized by init_early, including unloading crypto providers and cleaning up the network context.*/

static void
uninit_early(struct context *c)
{
    for (int j = 1; j < MAX_PARMS && c->options.providers.providers[j]; j++)
    {
        crypto_unload_provider(c->options.providers.names[j],
                               c->options.providers.providers[j]);
    }
    net_ctx_free(&c->net_ctx);
}

/**************************************************************************/
/**
 * OpenVPN's main init-run-cleanup loop.
 * @ingroup eventloop
 *
 * This function contains the two outer OpenVPN loops.  Its structure is
 * as follows:
 *  - Once-per-process initialization.
 *  - Outer loop, run at startup and then once per \c SIGHUP:
 *    - Level 1 initialization
 *    - Inner loop, run at startup and then once per \c SIGUSR1:
 *      - Call event loop function depending on client or server mode:
 *        - \c tunnel_point_to_point()
 *        - \c tunnel_server()
 *    - Level 1 cleanup
 *  - Once-per-process cleanup.
 *
 * @param argc - Commandline argument count.
 * @param argv - Commandline argument values.
 */
static int
openvpn_main(int argc, char *argv[])
{
    // This structure (struct context) holds all state-related information needed for OpenVPN.
    struct context c;

#if PEDANTIC
    fprintf(stderr, "Sorry, I was built with --enable-pedantic and I am incapable of doing any real work!\n");
    return 1;
#endif

#ifdef _WIN32
    SetConsoleOutputCP(CP_UTF8);
#endif

    CLEAR(c);

    /* signify first time for components which can
     * only be initialized once per program instantiation. */
    c.first_time = true;

    /* initialize program-wide statics */
    if (init_static())
    {
        /*
         * This loop is initially executed on startup and then
         * once per SIGHUP.
         */
        do
        {
            /* enter pre-initialization mode with regard to signal handling */
            pre_init_signal_catch();

            /* zero context struct but leave first_time member alone */
            context_clear_all_except_first_time(&c);

            /* static signal info object */
            c.sig = &siginfo_static;

            /* initialize garbage collector scoped to context object */
            gc_init(&c.gc);

            /* initialize environmental variable store */
            c.es = env_set_create(NULL);
#ifdef _WIN32
            set_win_sys_path_via_env(c.es);
#endif

#ifdef ENABLE_MANAGEMENT
            /* initialize management subsystem */
            init_management();
#endif

            /* initialize options to default state */
            init_options(&c.options, true);

            /* parse command line options, and read configuration file */
            parse_argv(&c.options, argc, argv, M_USAGE, OPT_P_DEFAULT, NULL, c.es);

#ifdef ENABLE_PLUGIN
            /* plugins may contribute options configuration */
            init_verb_mute(&c, IVM_LEVEL_1);
            init_plugins(&c);
            open_plugins(&c, true, OPENVPN_PLUGIN_INIT_PRE_CONFIG_PARSE);
#endif

            /* Early initialisation that need to happen before option
             * post processing and other early startup but after parsing */
            init_early(&c);

            /* set dev options */
            init_options_dev(&c.options);

            /* openssl print info? */
            if (print_openssl_info(&c.options))
            {
                break;
            }

            /* --genkey mode? */
            if (do_genkey(&c.options))
            {
                break;
            }

            /* tun/tap persist command? */
            if (do_persist_tuntap(&c.options, &c.net_ctx))
            {
                break;
            }

            /* sanity check on options */
            options_postprocess(&c.options, c.es);

            /* show all option settings */
            show_settings(&c.options);

            /* print version number */
            msg(M_INFO, "%s", title_string);
#ifdef _WIN32
            show_windows_version(M_INFO);
#endif
            show_library_versions(M_INFO);

            show_dco_version(M_INFO);

            /* misc stuff */
            pre_setup(&c.options);

            /* test crypto? */
            if (do_test_crypto(&c.options))
            {
                break;
            }

            /* Query passwords before becoming a daemon if we don't use the
             * management interface to get them. */
            if (!(c.options.management_flags & MF_QUERY_PASSWORDS))
            {
                init_query_passwords(&c);
            }

            /* become a daemon if --daemon */
            if (c.first_time)
            {
                c.did_we_daemonize = possibly_become_daemon(&c.options);
                write_pid_file(c.options.writepid, c.options.chroot_dir);
            }

#ifdef ENABLE_MANAGEMENT
            /* open management subsystem */
            if (!open_management(&c))
            {
                break;
            }
            /* query for passwords through management interface, if needed */
            if (c.options.management_flags & MF_QUERY_PASSWORDS)
            {
                init_query_passwords(&c);
            }
#endif

            /* set certain options as environmental variables */
            setenv_settings(c.es, &c.options);

            /* finish context init */
            context_init_1(&c);

            do
            {
                /* run tunnel depending on mode */
                switch (c.options.mode)
                {
                case MODE_POINT_TO_POINT:
                    tunnel_point_to_point(&c);
                    break;

                case MODE_SERVER:
                    tunnel_server(&c);
                    break;

                default:
                    ASSERT(0);
                }

                /* indicates first iteration -- has program-wide scope */
                c.first_time = false;

                /* any signals received? */
                if (IS_SIG(&c))
                {
                    print_signal(c.sig, NULL, M_INFO);
                }

                /* pass restart status to management subsystem */
                signal_restart_status(c.sig);
                /*SIGUSR1 loop: Restarts the VPN connection.*/
            } while (signal_reset(c.sig, SIGUSR1) == SIGUSR1);

            env_set_destroy(c.es);
            uninit_options(&c.options);
            gc_reset(&c.gc);
            uninit_early(&c);
            /*SIGHUP loop: Reloads the VPN configuration.*/
        } while (signal_reset(c.sig, SIGHUP) == SIGHUP);
    }

    context_gc_free(&c);

#ifdef ENABLE_MANAGEMENT
    /* close management interface */
    close_management();
#endif

    /* uninitialize program-wide statics */
    uninit_static();

    openvpn_exit(OPENVPN_EXIT_STATUS_GOOD); /* exit point */
    return 0;                               /* NOTREACHED */
}

/*

different main function implementations for Windows vs other operating systems.

When you run a program from the command line like:

openvpn --config server.conf

or

sudo systemctl start openvpn@myserver
[/usr/sbin/openvpn --config /etc/openvpn/server/myserver.conf]

*/
#ifdef _WIN32
int wmain(int argc, wchar_t *wargv[])
{
    char **argv;
    int ret;
    int i;

    if ((argv = calloc(argc + 1, sizeof(char *))) == NULL)
    {
        return 1;
    }

    for (i = 0; i < argc; i++)
    {
        int n = WideCharToMultiByte(CP_UTF8, 0, wargv[i], -1, NULL, 0, NULL, NULL);
        argv[i] = malloc(n);
        WideCharToMultiByte(CP_UTF8, 0, wargv[i], -1, argv[i], n, NULL, NULL);
    }

    ret = openvpn_main(argc, argv);

    for (i = 0; i < argc; i++)
    {
        free(argv[i]);
    }
    free(argv);

    return ret;
}
#else  /* ifdef _WIN32 */
int main(int argc, char *argv[])
{
    return openvpn_main(argc, argv);
}
#endif /* ifdef _WIN32 */
